generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -691,6 +698,124 @@ public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
 
   }
 
+  /**
+   * Identity transformation for integer types (both unsigned and signed, and their vectors) that
+   * double bitwise inverts an integer, producing the same integer as output. When performed,
+   * transforms an expression, e, such that:
+   *    e -> ~(~(expr)).
+   */
+  private class IdentityBitwiseNotNot extends AbstractIdentityTransformation {
+    private IdentityBitwiseNotNot() {
+      super(BasicType.allIntegerTypes(), false);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      assert BasicType.allIntegerTypes().contains(type);
+      // Invert once
+      Expr result = new UnaryExpr(new ParenExpr(applyIdentityFunction(expr, type,
+          constContext,
+          depth, fuzzer)), UnOp.BNEG);
+      // Invert again
+      result = new UnaryExpr(new ParenExpr(applyIdentityFunction(result, type, constContext,
+          depth, fuzzer)), UnOp.BNEG);
+      return identityConstructor(expr, result);
+    }
+  }
+
+  /**
+   * Identity transformation for integer types (both unsigned and signed, and their vectors) that
+   * ORs an integer with itself, producing the same integer as output. This identity requires
+   * expressions to be side effect free because the same mutated expression is evaluated twice.
+   * When performed, transforms an expression, e, such that:
+   *    e -> (e) | (e).
+   */
+  private class IdentityBitwiseOrSelf extends AbstractIdentityTransformation {
+    private IdentityBitwiseOrSelf() {
+      super(BasicType.allIntegerTypes(), true);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      assert BasicType.allIntegerTypes().contains(type);
+      // We use parentheses to prevent issues with order of operations in ternary expressions.
+      return identityConstructor(
+          expr,
+          new BinaryExpr(
+              new ParenExpr(applyIdentityFunction(expr.clone(), type, constContext, depth, fuzzer)),
+              new ParenExpr(applyIdentityFunction(expr.clone(), type, constContext, depth, fuzzer)),
+                  BinOp.BOR));
+    }
+  }
+  
+  /**
+   * Identity transformation for integer types (both unsigned and signed, and their vectors) that
+   * ORs an integer with zero, producing the same integer as output.
+   * When performed, transforms an expression, e, such that:
+   *    e -> (e) | (opaque 0).
+   */
+  private class IdentityBitwiseOrZero extends AbstractIdentityTransformation {
+    private IdentityBitwiseOrZero() {
+      super(BasicType.allIntegerTypes(), false);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      assert BasicType.allIntegerTypes().contains(type);
+      return applyBinaryIdentityFunction(
+          expr.clone(),
```

Comment:
From studying IdentityAddSubZero, I don't think you need to put the opaque zero in a ParenExpr (i.e. I believe sufficient parentheses should come with the opaque zero).

Can you check the various places you add parentheses to see whether they are required?

Reply:
You're correct - I just tested them and they aren't required. Thanks for noticing that.